roofit/roofitcore/inc/LinkDef.h

```diff
@@ -250,6 +250,13 @@
 #pragma link C++ class RooResolutionModel+ ;
 #pragma link C++ class RooTruthModel+ ;
 #pragma link C++ class RooProdPdf+ ;
+#pragma read sourceClass="RooProdPdf" targetClass="RooProdPdf" version="[4]" \
+  source="RooLinkedList _pdfNSetList" target="_pdfNSetList"                  \
+  code="{RooFIter iter = onfile._pdfNSetList.fwdIterator();                  \
+         while(auto nset = (RooArgSet*)iter.next()) {                        \
+           _pdfNSetList.emplace_back(*nset);                                 \
+         }                                                                   \
+         }";
```

Comment:
This could potentially leak. It's worth checking at least: You're creating a copy of the objects pointed to by the linked list. When the I/O operation is finished, the original objects are still around, and I'm not sure that anybody owns them. I hope that the linked list does, and will delete them once everything is deserialised, but I'm not sure.

At the same time, you cannot destroy them here in this loop, since they might be (co-)owned by another entity in the same file. Did you check what happens?

Reply:
Very good point, thanks for noticing! I solved this problem now by avoiding the copy. The `RooLinkedList` is now replaced with a `std::vector<std::unique_ptr<RooArgSet>>` instead of a `std::vector<RooArgSet>`. So the RooArgSet is not copied anymore, the ownership is only transferred from the `RooLinkedList` to the `unique_ptr`.

Fortunately we don't need to worry about co-owning, because as you also noticed from the RooProdPdf constructor the RooProdPdf object is responsible to call `Delete()` on the list.

Reply:
I think this is doing the right thing now. I don't think it's the case (since the member is private), but just be aware that there might be a file somewhere in the world where two objects see those pointers. Since you don't touch the pointer during the I/O operation, they will still have "shared ownership" after the I/O operation completes, but from then, the ProdPdf owns exclusively (via the `unique_ptr`), while the other object just observes.

That's probably all fine, because people observing the internal norm sets are asking for trouble. You could only achieve this via friend classes or other evil things.